File: route/bbs_w_set.py
from .tool.func import *

def bbs_w_set(bbs_num = ''):
    with get_db_connect() as conn:
        curs = conn.cursor()

        curs.execute(db_change('select set_data from bbs_set where set_id = ? and set_name = "bbs_name"'), [bbs_num])
        db_data = curs.fetchall()
        if not db_data:
            return redirect('/bbs/main')
        else:
            bbs_name = db_data[0][0]
        
        bbs_num_str = str(bbs_num)

        i_list = ['bbs_acl', 'bbs_edit_acl', 'bbs_comment_acl', 'bbs_view_acl', 'bbs_markup']

        if flask.request.method == 'POST':
            if admin_check(None, 'bbs_set (acl)') != 1:
                return re_error('/ban')
            else:
                for for_a in range(len(i_list)):
                    curs.execute(db_change("update bbs_set set set_data = ? where set_name = ? and set_id = ?"), [
                        flask.request.form.get(i_list[for_a], 'normal'),
                        i_list[for_a],
                        bbs_num
                    ])

                conn.commit()

                return redirect('/bbs/set/' + bbs_num_str)
        else:
            d_list = ['' for _ in range(0, len(i_list))]

            other_menu = []
            if admin_check() != 1:
                disable = 'disabled'
            else:
                disable = ''
                other_menu += [['bbs/delete/' + bbs_num_str, load_lang('delete')]]

            for for_a in range(len(i_list)):
                curs.execute(db_change('select set_data from bbs_set where set_name = ? and set_id = ?'), [i_list[for_a], bbs_num])
                sql_d = curs.fetchall()
                if sql_d:
                    d_list[for_a] = sql_d[0][0]
                else:
                    curs.execute(db_change('insert into bbs_set (set_name, set_code, set_id, set_data) values (?, "", ?, ?)'), [i_list[for_a], bbs_num, 'normal'])
                    d_list[for_a] = 'normal'

            conn.commit()

            acl_div = ['' for _ in range(0, len(i_list))]
            acl_list = get_acl_list()
            for for_a in range(0, len(i_list)):
                if for_a == 4:
                    acl_list = ['normal'] + get_init_set_list('markup')['list']

                for data_list in acl_list:
                    if data_list == d_list[for_a]:
                        check = 'selected="selected"'
                    else:
                        check = ''

                    acl_div[for_a] += '<option value="' + data_list + '" ' + check + '>' + (data_list if data_list != '' else 'normal') + '</option>'

            return easy_minify(flask.render_template(skin_check(),
                imp = [load_lang('bbs_set'), wiki_set(), wiki_custom(), wiki_css(['(' + bbs_name + ')', 0])],
                data = render_simple_set('''
                    <form method="post">
                        <hr class="main_hr">
                        <a href="/acl/TEST#exp">(''' + load_lang('reference') + ''')</a>
                        
                        <h2>''' + load_lang('acl') + '''</h2>
                        <h3>''' + load_lang('bbs_view_acl') + '''</h3>
                        <select ''' + disable + ''' name="bbs_view_acl">''' + acl_div[3] + '''</select>

                        <h4>''' + load_lang('bbs_acl') + '''</h4>
                        <select ''' + disable + ''' name="bbs_acl">''' + acl_div[0] + '''</select>

                        <h5>''' + load_lang('bbs_edit_acl') + '''</h5>
                        <select ''' + disable + ''' name="bbs_edit_acl">''' + acl_div[1] + '''</select>

                        <h5>''' + load_lang('bbs_comment_acl') + '''</h5>
                        <select ''' + disable + ''' name="bbs_comment_acl">''' + acl_div[2] + '''</select>

                        <h2>''' + load_lang('markup') + '''</h2>
                        <select ''' + disable + ''' name="bbs_markup">''' + acl_div[4] + '''</select>
                        
                        <hr class="main_hr">
                        <button id="opennamu_save_button" type="submit">''' + load_lang('save') + '''</button>
                    </form>
                '''),
                menu = [['bbs/w/' + bbs_num_str, load_lang('return')]] + other_menu
            ))

File: views/main_css/js/route/bbs_w_set.js
"use strict";

function opennamu_bbs_w_set_post() {
    let acl_set_list = [
        "bbs_view_acl",
        "bbs_acl",
        "bbs_edit_acl",
        "bbs_comment_acl",
        "bbs_markup",
        "bbs_name"
    ];

    const url = window.location.pathname;
    const url_split = url.split('/');

    let set_id = url_split[3];

    for(let for_a = 0; for_a < acl_set_list.length; for_a++) {
        let post_data = new FormData();
        post_data.append('data', document.getElementById('opennamu_' + acl_set_list[for_a]).value);
        
        fetch('/api/v2/bbs/set/' + set_id + '/' + acl_set_list[for_a], {
            method : 'PUT',
            body : post_data,
        }).then(function(res) {
            return res.json();
        }).then(function(data) {
            history.go(0);
        });
    }
}

function opennamu_bbs_w_set_lang(lang, set_name) {
    if(set_name === "bbs_markup") {
        return lang["markup"];
    } else {
        return lang[set_name];
    }
}

function opennamu_bbs_w_set_select(acl_set_list, acl_set_list_h, acl_list, lang) {
    let acl_set_html = '';

    for(let for_b = 0; for_b < acl_set_list.length; for_b++) {
        acl_set_html += '<h' + acl_set_list_h[for_b] + '>' + opennamu_bbs_w_set_lang(lang, acl_set_list[for_b]) + '</h' + acl_set_list_h[for_b] + '>';
        acl_set_html += '<select id="opennamu_' + acl_set_list[for_b] + '">';
        
        let select = '';
        for(let for_a = 0; for_a < acl_list.length; for_a++) {
            let acl_list_view = acl_list[for_a];
            acl_list_view = acl_list_view === "" ? "normal" : acl_list_view;

            select += '<option value="' + acl_list[for_a] + '">' + acl_list_view + '</option>';
        }

        acl_set_html += select;
        acl_set_html += '</select>';
    }

    return acl_set_html;
}

function opennamu_bbs_w_set() {
    const url = window.location.pathname;
    const url_split = url.split('/');

    let set_id = url_split[3];

    let acl_set_list = ["bbs_view_acl", "bbs_acl", "bbs_edit_acl", "bbs_comment_acl"];
    let acl_set_list_h = [2, 3, 4, 4];

    let markup_set_list = ["bbs_markup"];
    let markup_set_list_h = [2];
    
    let lang_str = 'save reference markup';
    for(let for_a = 0; for_a < acl_set_list.length; for_a++) {
        lang_str += ' ' + acl_set_list[for_a];
    }

    let lang_data = new FormData();
    lang_data.append('data', lang_str);

    let make_html = '';

    fetch('/api/v2/lang', {
        method : 'POST',
        body : lang_data,
    }).then(function(res) {
        return res.json();
    }).then(function(lang) {
        lang = lang["data"];

        fetch('/api/v2/list/acl/normal').then(function(res) {
            return res.json();
        }).then(function(acl_list) {
            acl_list = acl_list["data"];

            let acl_set_html = '<a href="/acl/TEST#exp">(' + lang['reference'] + ')</a>';
            acl_set_html += opennamu_bbs_w_set_select(acl_set_list, acl_set_list_h, acl_list, lang);

            make_html += acl_set_html;

            return fetch('/api/v2/list/markup');
        }).then(function(res) {
            return res.json();
        }).then(function(markup_list) {
            markup_list = markup_list["data"];

            make_html += opennamu_bbs_w_set_select(markup_set_list, markup_set_list_h, markup_list, lang);

            return;
        }).then(function() {
            document.getElementById('opennamu_bbs_w_set').innerHTML = renderSimpleSet('' +
                make_html +
                '<hr class="main_hr">' + 
                '<input id="opennamu_bbs_name">' +
                '<hr class="main_hr">' + 
                '<button onclick="opennamu_bbs_w_set_post();">' + lang['save'] + '</button>' +
            '');

            let total_set_list = [];
            total_set_list = total_set_list.concat(acl_set_list);
            total_set_list = total_set_list.concat(markup_set_list);

            for(let for_a = 0; for_a < total_set_list.length; for_a++) {
                fetch('/api/v2/bbs/set/' + set_id + '/' + total_set_list[for_a]).then(function(res) {
                    return res.json();
                }).then(function(data) {
                    data = data["data"][0][0];

                    let select_element = document.getElementById('opennamu_' + total_set_list[for_a]);
                    select_element.querySelector('option[value="' + data + '"]').selected = true;
                });
            }

            fetch('/api/v2/bbs/set/' + set_id + '/bbs_name').then(function(res) {
                return res.json();
            }).then(function(data) {
                data = data["data"][0][0];

                document.getElementById('opennamu_bbs_name').value = data;
            });
        });
    });
}